src/Components/GEOSldas_GridComp/GEOSlandassim_GridComp/clsm_ensupd_read_obs.F90

```diff
@@ -1548,6 +1548,436 @@ subroutine read_obs_sm_ASCAT(                                  &
+      ! Are we in the required assimilation window?
+      !e.g. Y2019/M07/D02/M01-ASCA-ASCSMO02-NA-5.0-20190702075700.000000000Z-20190702084627-1350204.bfr
+      str_date_time = tmpfname(40:53)
+      
+      read(str_date_time(1:4), *) date_time_tmp%year
+      read(str_date_time(5:6), *) date_time_tmp%month
+      read(str_date_time(7:8), *) date_time_tmp%day
+      read(str_date_time(9:10), *) date_time_tmp%hour
+      read(str_date_time(11:12), *) date_time_tmp%min
+      read(str_date_time(13:14), *) date_time_tmp%sec
+  
+      if ( datetime_lt_refdatetime( date_time_low_fname, date_time_tmp ) .and.        &
+            datetime_le_refdatetime( date_time_tmp, date_time_up )) then 
+      N_tmp = N_tmp + 1
+      ! Remove the '/D03/' from the directory part as using "read_obs_SMAP_fnames" 
+      ind = index(tmpfname, "/D")
+      
+      if (ind > 0) then
+         tmpfname2 = tmpfname(1:ind) // tmpfname(ind+5:)
+      end if
+      tmpfnames(N_tmp) = trim(this_obs_param%path) // '/' // trim(tmpfname2)
+      end if
+     end do
+   
+    fnames = tmpfnames(1:N_tmp)
+    N_files = N_tmp
+
+    if (N_files>0) then
+
+
+       ! read and process data if files are found
+       allocate(tmp1_lon(max_rec))
+       allocate(tmp1_lat(max_rec))
+       allocate(tmp1_obs(max_rec))
+       allocate(tmp1_jtime(max_rec))
+       
+       call MTINFO( trim(this_obs_param%path) // '/BUFR_mastertable/', 51, 52)
+       ! file loop
+       N_tmp = 0
+       do kk = 1,N_files
+
+          ! open on bufr file
+          call closbf(lnbufr)
+          open(lnbufr, file=trim(fnames(kk)), action='read',form='unformatted')
+          call openbf(lnbufr,'SEC3', lnbufr)
+          call MTINFO( trim(this_obs_param%path) // '/BUFR_mastertable/', 51, 52)
+          call datelen(10)
+         
+          msg_report: do while(ireadmg(lnbufr,subset,idate) ==0)
+            loop_report: do while(ireadsb(lnbufr) == 0)
+            ! extract sensing time information 
+                 call ufbint(lnbufr,tmp_time,6,1,iret,'YEAR MNTH DAYS HOUR MINU SECO')
+                 date_time_tmp.year = int(tmp_time(1))
+                 date_time_tmp.month = int(tmp_time(2)) 
+                 date_time_tmp.day = int(tmp_time(3))
+                 date_time_tmp.hour = int(tmp_time(4))
+                 date_time_tmp.min = int(tmp_time(5))
+                 date_time_tmp.sec = int(tmp_time(6))
+
+                 ! skip if record outside of current assim window
+                 if ( datetime_lt_refdatetime( date_time_tmp, date_time_low ) .and.        &
+                      datetime_le_refdatetime( date_time_up, date_time_tmp )) cycle loop_report
+
+
+                  call ufbint(lnbufr,tmp_data,7,1,iret,'SSOM DOMO SMPF SMCF ALFR TPCX IWFR') 
+
+                
+                 ! skip if record contain no valid soil moisture value
+!                 call ufbint(lnbufr,tmp_data,1,1,iret,'SSOM')
+                 if(tmp_data(1) > 100. .or. tmp_data(1) < 0.) cycle loop_report
```

Comment:
@amfox37, @gmao-qliu, @mathomp4: I'm not 100%, but it looks like these "while" loops read each obs+metadata separately.  If this is indeed the case, is there a way to read the entire file in one fell swoop and then apply the time window etc when we have everything in memory?

Reply:
Yes, I'd already tried consolidating the QC/flag reads (sorry for the nasty commented out lines in the draft) which didn't seem to impact things much, but could try consolidating the whole thing and then applying a QC loop.